### PR TITLE
Feature/favourites

### PR DESCRIPTION
**Pull request by stepan-zubkov**
New feature: **favourites product!**
Added:
- M2M between User and Product
- Index view for favourites
- Ajax endpoint for adding product to favourites
- Functionality for "В избранное" button in cart index page
- Tests for app

### CODE REVIEW

File: butterfly/favourites/views.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+from butterfly.products.models import Product
+from butterfly.cart.models import CartItem, Cart
+
+
+@login_required
+def add_product(request: HttpRequest):
+    if request.method == "POST" and request.is_ajax():
+        product_id = request.POST.get("product_id", "")
+        if not product_id.isdigit():
+            raise Http404()
```
Comment (kirillzhosul) [inline butterfly/favourites/views.py:14]: It is not 404, rather 400 (bad request)

File: butterfly/favourites/views.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        CartItem.objects.filter(product=product, cart=cart).delete()
+
+        return JsonResponse(
+            {
+                "product": {"name": product.name},
+                "cart_total_price": cart.get_total_price(),
+            }
+        )
+
+    raise Http404()
```
Comment (kirillzhosul) [inline butterfly/favourites/views.py:32]: It is not 404, rather than 405 (method not allowed)

File: butterfly/favourites/views.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        user = request.user
+        if not user.favourites.filter(id=product_id):
+            user.favourites.add(product)
+
+        cart = Cart.get_cart(request)
+        CartItem.objects.filter(product=product, cart=cart).delete()
+
+        return JsonResponse(
+            {
+                "product": {"name": product.name},
```
Comment (kirillzhosul) [inline butterfly/favourites/views.py:27]: There is any ID key?

File: butterfly/static/js/favourites.js [hunk 4]
```diff
@@ -0,0 +1,28 @@
+
+function addProductToFavourites(product_id) {
+    var this_script = $('script[src*="/static/js/favourites.js"]');
+    var url = this_script.attr('data-add-product-url');
+    var csrf = this_script.attr('data-csrf');
+
```
Comment (kirillzhosul) [inline butterfly/static/js/favourites.js:6]: Why not to use consts here?

**Review by kirillzhosul — APPROVED** [review]

File: butterfly/favourites/views.py (hunk 3 above)
Comment (stepan-zubkov) [reply butterfly/favourites/views.py:27]: product name is used for finding product on cart page by name (ID isn't displaying on cart page)

File: butterfly/static/js/favourites.js (hunk 4 above)
Comment (stepan-zubkov) [reply butterfly/static/js/favourites.js:6]: Maybe

File: butterfly/favourites/views.py (hunk 2 above)
Comment (stepan-zubkov) [reply butterfly/favourites/views.py:32]: Ok

File: butterfly/favourites/views.py (hunk 1 above)
Comment (stepan-zubkov) [reply butterfly/favourites/views.py:14]: Ok

**Pull request merged**